target/getmail/getmail-service.sh

```diff
@@ -0,0 +1,47 @@
+#!/bin/bash
+
+# shellcheck source=../scripts/helpers/log.sh
+source /usr/local/bin/helpers/log.sh
+
+# Directory, where "oldmail" files are stored.
+# getmail stores its state - its "memory" of what it has seen in your POP/IMAP account - in the oldmail files.
+GETMAIL_DIR=/var/lib/getmail
+
+# Kill all child processes on EXIT.
+# Otherwise 'supervisorctl restart getmail' leads to zombie 'sleep' processes.
+trap 'pkill --parent $$' EXIT
+
+function _syslog_error() {
+  logger --priority mail.err --tag getmail "${1}"
+}
+
+function _stopService() {
+  _syslog_error "Stopping service"
+  exec supervisorctl stop getmail
+}
+
+# Verify the correct value for GETMAIL_POLL. Valid are any numbers greater than 0.
+if ! [[ ${GETMAIL_POLL} =~ ^[0-9]+$ && ${GETMAIL_POLL} -gt 0 ]]; then
+  _syslog_error "Invalid value for GETMAIL_POLL: ${GETMAIL_POLL}"
+  _stopService
+fi
+
+# If no matching filenames are found, and the shell option nullglob is disabled, the word is left unchanged.
+# If the nullglob option is set, and no matches are found, the word is removed.
+shopt -s nullglob
+
+# Run each getmailrc periodically.
+while :; do
+  for RC_FILE in /etc/getmailrc.d/*; do
+    _log 'debug' "Processing ${RC_FILE}"
+    getmail --getmaildir "${GETMAIL_DIR}" --rcfile "${RC_FILE}"
+  done
+
+  # Stop service if no configuration is found.
+  if [[ -z ${RC_FILE} ]]; then
+    _syslog_error 'No configuration found'
+    _stopService
+  fi
```

Comment:
I quickly understood what you're doing here: the `for` loop will not set the `RC_FILE` variable if no config files exist, due to the `nullglob`. This is non-trivial, though, as you'd normally not expect the variable `RC_FILE` to live longer than the `for` loop; Bash is doing Bash things here...

I'd rather check first with something like this: https://superuser.com/a/1361548; of course, use `[[ -n ...` instead of `test -n`.

Reply:
> This is non-trivial, though, as you'd normally not expect the variable RC_FILE to live longer than the for loop; Bash is doing Bash things here...

I disagree, why would some one think a variable to magically disappears only because used in a loop?

> This is non-trivial, though, as you'd normally not expect the variable RC_FILE to live longer than the for loop

I do 😆

Reply:
> > This is non-trivial, though, as you'd normally not expect the variable RC_FILE to live longer than the for loop; Bash is doing Bash things here...
> 
> I disagree, why would someone assume a variable to magically disappear just because it was used in a loop?

I heavily disagree on this take:

```
int main()
{
  for (int i = 0; i < 10; ++i)
  {
    // do something with i
  }

  printf("%d", i);
}
```

This is not a valid program in C.

```rust
fn main()
{
  for i in 0..10
  {
    // do something with I
  }

  println("{i}");
}
```

This is not a valid program in Rust.

```java
public static void main(String[] args)
{
  for(int i=0; i<10; i++)
  {
      // do something with i
  }

  System.out.println(i);
}
```

This is not a valid program in Java.

```go
package main

import "fmt"

func main()
{
  for i := 0; i < 10; i++
  {
		sum += i
  }

  fmt.Println(i)
}
```

This is not a valid program in Go.

I could go on and on with this.

> > This is non-trivial, though, as you'd normally not expect the variable RC_FILE to live longer than the for loop
> 
> I do 😆 

This is probably because you write Bash very well and know it's ins and outs.

> I'd like to keep it as it is. As always, there are multiple ways to achieve something and everyone have it's on preferences how to do it 👍 
>
> Not really relevant, but my solution is more performant, as it does not invoke a subshell / external tool. See also my post here https://github.com/docker-mailserver/docker-mailserver/pull/4156#discussion_r1719059578

Let's be real, whether this takes 5ms more is not reall relevant here. My point is reall more about whether this is intuitive.

This is fine for me though. I don't think anyone is bothered, and those that do read our code are probably sufficient in Bash to understand what's going on here.

Reply:
> Let's be real, whether this takes 5ms ..

Absolutely. As I said "Not really relevant" here.

---

For completeness:

This is a valid PHP script:

```php
<?php
for ($i = 1; $i <= 10; $i++) {
    echo "$i\n";
}
echo $i;
```

This is valid JS script:

```js
for (var i = 1; i <= 10; i++) {
    console.log(i);
}
console.log(i);
```

This is valid python program

```python
for i in range(1, 11):
    print(i)

print(i)
```

This is valid ruby program:

```ruby
for i in 1..10
    puts i
end

puts i
```

This is valid perl script:

```perl
for my $i (1..10) {
    print "$i\n";
}

print "$i\n";
```

Reply:
It seems we have an interesting difference between compiled and interpreted languages here (the Go example does not count; you explicitly declare `i` _before_ the loop)..

As I said, not a blocker from my side.

Reply:
Just for clarity, the behaviour being described with the interpreted languages is known as variable hoisting.

JS should not use `var` these days, you'd use `let` which will properly bind the scope (~~no var hoisting~~)... and you'll find the result after the loop is then `null`.

Variable hoisting is typically an anti-pattern as it can cause unexpected bugs if you're not careful. (_**EDIT:** [`let` does hoist, but does not initialize](https://dev.to/shubhsharma19/4-habits-to-avoid-bugs-related-to-hoisting-in-your-javascript-code-4i3i#comment-2ef17), which is why `var` is considered `undefined` when referenced whilst with `let` it'll fail with an error_)